src/components/Cluster/ClusterDetail/ClusterApps.tsx

```diff
@@ -0,0 +1,356 @@
+import UserInstalledApps from './UserInstalledApps/UserInstalledApps';
+
+function formatAppVersion(
+  release: IRelease,
+  appMeta: AppConstants.IAppMetaApp
+) {
+  const { name, version } = appMeta;
+
+  if (name === 'kubernetes' && release.k8sVersionEOLDate) {
+    const { isEol } = getKubernetesReleaseEOLStatus(release.k8sVersionEOLDate);
+    if (isEol) {
+      return `${version} ${Constants.APP_VERSION_EOL_SUFFIX}`;
+    }
+  }
+
+  if (!version) {
+    return <NotAvailable />;
+  }
+
+  return version;
+}
+
+const SmallHeading = styled.h6`
+  font-size: 12px;
+  font-weight: 700;
+  text-transform: uppercase;
+  letter-spacing: 2px;
+`;
+
+const BrowseButtonContainer = styled.div`
+  display: flex;
+  align-items: center;
+  margin: ${({ theme }) => theme.spacingPx}px 0
+    ${({ theme }) => theme.spacingPx * 2}px;
+`;
+
+const BrowseButton = styled(Button)`
+  margin: 0;
+`;
+
+const Disclaimer = styled.p`
+  margin: 0 0 20px;
+  line-height: 1.2;
+`;
+
+const PreinstalledApps = styled.div`
+  display: flex;
+
+  & > div {
+    flex: 1;
+    margin-right: 15px;
+
+    &:last-child {
+      margin-right: 0px;
+    }
+  }
+`;
+
+interface IClusterAppsProps {
+  clusterId: string;
+  hasOptionalIngress?: boolean;
+  showInstalledAppsBlock?: boolean;
+  release?: IRelease;
+  installedApps?: IInstalledApp[];
+}
+
+const ClusterApps: React.FC<IClusterAppsProps> = ({
+  clusterId,
+  hasOptionalIngress,
+  showInstalledAppsBlock,
+  release,
+  installedApps,
+}) => {
+  const [detailsModalIsVisible, setDetailsModalIsVisible] = useState(false);
+  const [detailsModalAppName, setDetailsModalAppName] = useState('');
+
+  const appToDisplay = useMemo(() => {
+    return installedApps?.find(
+      (app) => app.metadata.name === detailsModalAppName
+    );
+  }, [installedApps, detailsModalAppName]);
+
+  useLayoutEffect(() => {
+    if (!detailsModalIsVisible || appToDisplay) return;
+
+    new FlashMessage(
+      'The app you were looking at was deleted from the cluster by someone else.',
+      messageType.ERROR,
+      messageTTL.LONG
+    );
+
+    setDetailsModalIsVisible(false);
+    setDetailsModalAppName('');
+  }, [detailsModalIsVisible, appToDisplay, installedApps]);
+
+  // This makes a list of apps that are installed as part of the cluster creation.
+  // It combines information from the release endpoint with the latest info
+  // coming from App CRs.
+  const preInstalledApps = useMemo(() => {
+    const displayApps: Record<
+      string,
+      Record<string, AppConstants.IAppMetaApp>
+    > = {
+      essentials: {},
+      management: {},
+      ingress: {},
+    };
+
+    if (release?.components) {
+      for (const { name, version } of release.components) {
+        if (!AppConstants.appMetas.hasOwnProperty(name)) continue;
+
+        let appMeta = AppConstants.appMetas[name];
+        if (typeof appMeta === 'function') {
+          appMeta = appMeta(version);
+        }
```

Comment:
`AppConstants.appMetas.hasOwnProperty(name)` is just for checking if the property exists.

`typeof appMeta === 'function'` for checking if the property is a function, and executing it if it is. This can also be a function because some apps have custom logic depending on the version used.

https://github.com/giantswarm/happa/blob/119e123e7705aa79fc7ff355d2eb8a20b54b7296/src/shared/constants/apps.ts#L34